Add insurance buffer

kumar-ish opened the pull request:
## Motivation

As per the Tracer Perpetual Swaps whitepaper, the insurance pool for a market is composed of two accounts: the system-owned buffer insurance account, and the public insurance account. This pull request implements the system-owned buffer insurance account.

## Changes
* Add insurance buffer account

* Change respective functions to account for the insurance buffer

Note: Scope of PR changed so as to not include the insurance withdrawal fee, as aspects of it are still under research and development

kumar-ish committed:
Add missing NatSpecs for TracerPerpetualSwaps constructor

kumar-ish committed:
Buffer commit

kumar-ish committed:
Merge branch 'contracts-v2' of https://github.com/tracer-protocol/tracer-protocol into insurance-buffer

kumar-ish committed:
Add insurance buffer account

* Add insurance buffer account, the system-owned insurance account

* Add fees to insurance pool withdrawals

kumar-ish committed:
Lint

kumar-ish committed:
Remove insurance withdrawal fee

Making an appearance soonTM

greenbookwebb commented on contracts/Insurance.sol:
```diff
@@ -68,11 +69,11 @@ contract Insurance is IInsurance, Ownable, SafetyWithdraw {
         InsurancePoolToken poolToken = InsurancePoolToken(token);
 
-        // tokens to mint = (pool token supply / collateral holdings) * collaterael amount to stake
+        // tokens to mint = (pool token supply / collateral holdings) * collateral amount to stake
```
How are we defining collateral holdings? Is collateral holdings just collateral amount (public), or is it collateral amount (public) + buffer? 

It should be the second one

CalabashSquash replied:
I think, at the moment, collateralAmount is just the public pool. However, it could probably be renamed to publicCollateralAmount.

greenbookwebb commented on contracts/Insurance.sol:
```diff
@@ -99,21 +100,26 @@ contract Insurance is IInsurance, Ownable, SafetyWithdraw {
         uint256 wadTokensToSend =
             LibInsurance.calcWithdrawAmount(
```
Just confiming that this function only allows you to withdraw from publc (collateralAmount), and not from bufferAmount.

CalabashSquash replied:
Yeah, I think `calcWithdrawAmount` should only take public collat amount, as opposed to public + buffer, right?

kumar-ish replied:
Ooh, you're definitely right; fixed now

greenbookwebb commented on contracts/Insurance.sol:
```diff
@@ -124,7 +130,12 @@ contract Insurance is IInsurance, Ownable, SafetyWithdraw {
         if (quote > 0) {
             tracer.withdraw(uint256(quote));
-            collateralAmount = collateralAmount + uint256(quote);
+            if (collateralAmount > 0) {
+                collateralAmount = collateralAmount + uint256(quote);
+            } else {
+                // Pay to buffer if nothing in public insurance
+                bufferAmount = bufferAmount + uint256(quote);
+            }
```
Some of this is true, some is incorrect. 

Insurance funding rate payments are paid proportionally to the two pools. e.g. If insurance funding payment = 25 Quote, Insurance buffer account has 10 units of quote already in the pool (10% of total), public account has 90 units of quote (90% of total), Insurance buffer account increases by 25*0.1, and public account increases by 25*0.9. 

BUT, if neither pool has any quote collateral in their account (like maybe at genesis), it will go 100% to insurance buffer account.

kumar-ish replied:
Yep, that makes sense; will fix this up

CalabashSquash replied:
Why do you say 250.1 and 250.9? Where have these numbers come from? EDIT: this is a markdown formatting thing lol

kumar-ish replied:
Yep, that makes sense; can fix

CalabashSquash replied:
According to @kumar-ish , the buffer account only gets paid when
a) The public account is empty
b) People pay withdrawal fees

But you're saying it's proportional to pool holdings.

If your statement is correct, should this split be done as soon as insurance funding rate payments are made, or is it OK to make this calculation when `updatePoolAmount(...)` is called?

kumar-ish replied:
This should be fixed now; also, discussed in-person with Ben and it should be okay to make the calculation when `updatePoolAmount()` is called

CalabashSquash replied:
Yep it is, because `updatePoolAmount()` is called on any change to insurance pool balance. I was worried the ratios could be "outdated" by the time it is called, but this is not the case.

greenbookwebb commented on contracts/Insurance.sol:
```diff
@@ -199,7 +214,7 @@ contract Insurance is IInsurance, Ownable, SafetyWithdraw {
         uint256 ratio =
             PRBMathUD60x18.div(
-                getPoolTarget() - collateralAmount,
+                getPoolTarget() - collateralAmount - bufferAmount,
```
unsure what this does.

kumar-ish replied:
This ratio is `(insurance_fund_target - insurance_fund_holdings)/(leveraged_notional_value)` in the formula

kumar-ish replied:
`collateralAmount` is the amount of quote in the public insurance pool and `bufferAmount` is the amount of quote in the buffer insurance pool

CalabashSquash replied:
This makes the calculation for funding rate equal to

`36523 * (10^11) * [(target - public collat amount - buffer collat amount) / leveraged notional value]`

greenbookwebb replied:
What's this for though? Insurance Funding Rate?

kumar-ish replied:
Yep, that's right

kumar-ish committed:
Fix insurance payments to public/buffer pool

kumar-ish committed:
Fix calcWithdrawAmount usage to not include buffer in collat holdings

& remove redundant cast

CalabashSquash commented on contracts/Insurance.sol:
```diff
@@ -142,24 +163,28 @@ contract Insurance is IInsurance, Ownable, SafetyWithdraw {
-            amount = collateralAmount - (10**18);
-            // Use new amount to compute difference again.
-            difference = collateralAmount - amount;
+            bufferAmount = 10**18;
```
I think this is wrong. You can't just arbitrarily set bufferAmount to 1. Where is this token coming from?

kumar-ish replied:
This is making the making sure there is at least one token in the insurance pool, but it should probably be a part of the public pool. I'm gonna speak to Webb about this and see how it should be specc'd

greenbookwebb replied:
yep, as discussed, get rid - no need for this. Only for public account

kumar-ish replied:
Fixed (I think)

kumar-ish committed:
Fix deposit to only mint as a proportion of public pool holdings

greenbookwebb commented on contracts/Insurance.sol:
```diff
@@ -199,7 +224,7 @@ contract Insurance is IInsurance, Ownable, SafetyWithdraw {
             PRBMathUD60x18.div(
-                getPoolTarget() - collateralAmount,
+                getPoolTarget() - collateralAmount - bufferAmount,
                 levNotionalValue
```
If `levNotionalValue == 0`, this will throw. This can happen naturally, or happen when `maxlev == 1`

greenbookwebb replied:
@BenjaminPatch @kumar-ish

kumar-ish replied:
There's a check for `levNotionalValue` on line 224; following block of code:
```
    if (levNotionalValue <= 0) {
        return 0;
    }
```

kumar-ish committed:
Fix drainPool to properly drain public insurance pool such that there's always at least one token left

kumar-ish committed:
Rename collateralAmount and bufferAmount to better reflect their purposes

kumar-ish committed:
Fix drainPool logic

kumar-ish committed:
Fix drainPool logic so that public collateral is capped

CalabashSquash commented on contracts/Insurance.sol:
```diff
@@ -137,29 +161,55 @@ contract Insurance is IInsurance, Ownable, SafetyWithdraw {
+            bufferCollateralAmount = 0;
+        } else if (amount > bufferCollateralAmount) {
+            uint256 leftAfterDrain =
+                amount - bufferCollateralAmount - publicCollateralAmount;
```
Consider the example where
`publicCollateralAmount = 10`
`bufferCollateralAmount = 20`
`amount = 25`

then `amount > bufferCollateralAmount == true`
so `leftAfterDrain = 25 - 20 - 10` so underflow would occur.

It should make `leftAfterDrain = amount - bufferCollateralAmount`, then check if `amount > public`, if so drain all but 1. Then `amount = amountTakenFromBuffer + amountTakenFromPublic`, and set bufferCollateralAmount to =, and amountTakenFromPublic to be whatever it should be (min 1 token)

kumar-ish replied:
Yes, you're right -- I believe the logic should be all _patched_ up now. I did some [tests](https://gist.github.com/kumar-ish/23ff05e1cdf52be6826e0f7ee4f500c0) in Python too, but feel free to attack it with edge cases

kumar-ish committed:
Fix drainPool edge cases

greenbookwebb commented on contracts/Insurance.sol:
```diff
@@ -68,17 +69,17 @@ contract Insurance is IInsurance, Ownable, SafetyWithdraw {
         // mint pool tokens, hold collateral tokens
         poolToken.mint(msg.sender, tokensToMint);
-        collateralAmount = collateralAmount + wadAmount;
+        publicCollateralAmount = publicCollateralAmount + wadAmount;
```
What is wadAmount? Is it the collateral that is being deposited into the pool when a user mints pool tokens?

kumar-ish replied:
Yes, that's right

greenbookwebb commented on contracts/Insurance.sol:
```diff
@@ -121,10 +123,32 @@ contract Insurance is IInsurance, Ownable, SafetyWithdraw {
-        if (quote > 0) {
-            tracer.withdraw(uint256(quote));
-            collateralAmount = collateralAmount + uint256(quote);
+        uint256 quote =
+            uint256((tracer.getBalance(address(this))).position.quote);
+
+        tracer.withdraw(quote);
+
+        if (publicCollateralAmount > 0) {
+            // Amount to pay to public is the ratio of public collateral amount to total funds
+            uint256 payToCollateral =
+                PRBMathUD60x18.mul(
+                    quote,
+                    PRBMathUD60x18.div(
+                        publicCollateralAmount,
+                        publicCollateralAmount + bufferCollateralAmount
+                    )
+                );
+
+            publicCollateralAmount = publicCollateralAmount + payToCollateral;
+
+            // Amount to pay to buffer is the remainder
+            bufferCollateralAmount =
+                bufferCollateralAmount +
+                quote -
+                payToCollateral;
+        } else {
+            // Pay to buffer if nothing in public insurance
+            bufferCollateralAmount = bufferCollateralAmount + quote;
```
This doesn't look right. At the time a user withdraws from the public fund, the withdrawal fee goes entirely to the insurance buffer account.

kumar-ish replied:
This is for updating the amount of buffer and public collateral tokens in the insurance pool. In order to optimise for gas, all insurance funding payments are made in the `TracerPerpetualSwaps` contract; when you actually need to deposit or withdraw from the insurance pool, you need access to those collateral tokens, so `updatePoolAmount` gets them from the perp swaps contract (which is why `tracer.withdraw(quote);` is called at the beginning). 

```
            // Pay to buffer if nothing in public insurance
```
is saying that if no one has deposited collateral into the insurance account, then any amount that the Tracer has paid out to the insurance contract (from insurance funding rate payments) belongs to the buffer account. If someone has (i.e. the if-statement branch, where the collateral in the insurance contract is greater than 0), you pay proportionally to the buffer and public account.

The withdraw fee is still under discussion so I decided to leave it out of scope for this PR.

greenbookwebb commented on contracts/Insurance.sol:
```diff
@@ -137,29 +161,46 @@ contract Insurance is IInsurance, Ownable, SafetyWithdraw {
-        if (amount > collateralAmount) {
-            amount = collateralAmount;
-        }
+        if (amount >= poolHoldings) {
+            // If public collateral left after draining is less than 1 token, we want to keep it at 1 token
+            if (publicCollateralAmount > 10**18) {
+                // Leave 1 token for the public pool
+                amount = poolHoldings - 10**18;
+                publicCollateralAmount = 10**18;
+            } else {
+                amount = bufferCollateralAmount;
+            }
+
+            // Drain buffer
+            bufferCollateralAmount = 0;
+        } else if (amount > bufferCollateralAmount) {
+            if (publicCollateralAmount < 10**18) {
+                // If there's not enough public collateral for there to be 1 token, cap amount being drained at the buffer
+                amount = bufferCollateralAmount;
+            } else if (poolHoldings - amount < 10**18) {
+                // If the amount of collateral left in the public insurance would be less than 1 token, cap amount being drained
+                // from the public insurance such that 1 token is left in the public buffer
+                amount =
+                    bufferCollateralAmount +
+                    publicCollateralAmount - 
+                    10**18;
+                publicCollateralAmount = 10**18;
+            } else {
+                // Take out what you need from the public pool; there's enough for there to be >= 1 token left
+                publicCollateralAmount = poolHoldings - amount;
+            }
```
this looks correct, but there's a lot going on here so I won't confirm - @BenjaminPatch to confirm

CalabashSquash replied:
There is a lot going on but there are a decent amount of edge cases that need checking. I'm confident that this is good.

greenbookwebb reviewed (changes requested):
Most of this looks good. Just got to fix up that bit with the withdrawal fee going entirely to the insurance buffer account.

CalabashSquash commented on contracts/Insurance.sol:
```diff
@@ -121,10 +123,32 @@ contract Insurance is IInsurance, Ownable, SafetyWithdraw {
+
+        tracer.withdraw(quote);
+
+        if (publicCollateralAmount > 0) {
```
Just confirming this is correct, and shouldn't be `bufferCollateralAmount`.

kumar-ish replied:
Yes; only when someone has deposited into the public insurance account do funds start accruing proportionally to the buffer and public pools; otherwise, they all go to the public pool

kumar-ish committed:
buffer testing commit

kumar-ish committed:
Pull upstream changes

kumar-ish committed:
Refactor Insurance to have getPoolHoldings()

kumar-ish committed:
Remove unnecessary addition in Insurance.sol

kumar-ish committed:
Add getters for poolAmount variables

kumar-ish committed:
buffer commit

kumar-ish committed:
Merge branch 'contracts-v2' of https://github.com/tracer-protocol/tracer-protocol into HEAD

kumar-ish committed:
Fix drainPool public collateral withdrawal

Previously returning amount left in pool overall; now returning how much left in public

kumar-ish committed:
Refactor insurance contract and add some documentation

kumar-ish committed:
Add tests for insurance buffer

kumar-ish committed:
Fix merge mistake in perp swap constructor NatSpec

kumar-ish committed:
dFix drainPool public collateral withdrawal #2

Bad test case now fixed

raymogg committed:
Merge branch 'contracts-v2' of https://github.com/tracer-protocol/tracer-protocol into insurance-buffer
